### ISSUE TITLE
 [Feature] Node mirroring for GPU-memory reduction

### ISSUE THREAD

**Issue by MarvinTeichmann** [tensorflow/tensorflow#11171]
In the paper [Training Deep Nets with Sublinear Memory Cost](https://arxiv.org/abs/1604.06174) Chen et al. introduced a very good idea to greatly reduce GPU memory requirements. 

The idea bowls down to discarding the output of some nodes during the forward pass and recompute those values when they are needed again in the backward pass. Only the output of some key ops is kept in memory. During back-prop all forward computation from those key nodes are redone. They also describe how this can be implemented in a graph based computation model by mirroring non-key ops. This is depicted in figure 3 (see below).

Performing this kind of graph manipulation in [MxNet](https://github.com/dmlc/mxnet) is quite easy and I have played around with this myself. I am able to reduce the memory cost of a SotA segmentation model from `13504 MB` to `3382 MB` for the cost of about ~40% increase in computational time. (Given that we have plenty 1080 TI and view P100 GPUs, I am very happy to pay this cost).

For me as deep learning researcher this is a totally awesome killer feature. In most of my experiments I am limited by the amount of available GPU memory. Doing node mirroring allows me to try a whole bunch of new stuff, I was always wanting to do. 

1. Is anything like this planned to be implemented in Tensorflow any time soon?
2. In the current API, is there already a way to build and / or manipulate the computational graph to perform node mirroring (like in figure 3)?

![](https://i.imgur.com/CnkUzwJ.png)

Regarding question two, I don't mind if it gets messy. Copying some nodes inside the graph is possible in tensorflow. Gradient flow can also be stopped for the first copy. What is missing is to utilize the second node for gradient computation. I don't know how I can archive this using the python API in tensorflow. Any ideas with this?

**Comment by skye** [comment]
@prb12 can you comment or redirect? Thanks.

**Comment by prb12** [comment]
I'm not really the best person to answer this, but yes we are aware of this technique and similar things crop up in various places in TensorFlow:

- XLA has an optimization pass that does selective rematerialization of HLO expressions when necessary to fit in a memory budget `tensorflow/tensorflow/compiler/xla/service/hlo_rematerialization.h`
- Grappler has a rewriting pass that does this: `tensorflow/tensorflow/core/grappler/optimizers/memory_optimizer.cc`
- Not strictly speaking recomputation, but the dynamic RNN classes support a 'swap_memory` flag that temporarily copies intermediate values required for backprop from the GPU back to host memory to allow very long sequence lengths without running out of GPU memory.

@benoitsteiner is the best person to answer Grappler questions, @ebrevdo for the RNN code, and @eliben or @meheffernan for the XLA work.

**Comment by eliben** [comment]
@sanjoy

**Comment by MarvinTeichmann** [comment]
Thanks for the answers. I had a short look through the gradient computation code in tensorflow. By now I am quite confident that node-mirroring can be be (fully) implemented in python. The whole think bowls done to manipulating the code which produces the [gradient computation subgraph](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/ops/gradients_impl.py#L488-L516).

At the moment the intermediate results of each tensor are saved in GPU memory (if gradients are computed). This is an *incredible* wasteful behavior. I think this is the main reason why tensorflow is so extremly hungry for GPU memory. 

As an example, I have recently implemented a (gaussian) crf model in tensorflow. When performing back-propagation through the CRF the memory requirement went up from `320MB` to `10.45GB`, compared to plain inference. The issue in this case is, that the different computation along the line of  `tf.exp ( tf.reduce_sum(   -0.5 * tf.square( f_i - f_j ) * M ) )` are done very often. For each minor step in those kind of computations (i.e. `-0.5 * sqr_diff` ) the intermediate output is kept in memory.  The time it take to run the entire core of the CRF is only `347 μs`. This is negligible, compared to the rest of the model. 

So the bottom line of this example is: Recomputation during backward pass could save 10GB for a negligible amount of extra computation time.

**Comment by MarvinTeichmann** [comment]
Also, I had a short look at the other memory optimization projects mentioned. XLA and grapper implement good ideas. However those projects aim at performing memory optimization fully automated. I feel for advanced users it will be much more valuable to give them options for manual memory optimization. I think many users know the bottleneck of their model and manual optimization can archive much better results then automated memory optimization projects. [See the crf example].

The swap memory used for RNNs is a pretty cool idea. This can be used as an alternative to the recomputation approach. I think it would be great to have a function which could perform `swap_memory` on arbitrary (i.e. non-rnn) tensors.  There are long chain models even outside the rnn world and this can be very useful for those as well. [*Btw. could someone point me to the line where swap_memory is implemented? I am not familiar with the rnn code. What I have seen so far most functions just pass this parameter on. What would it take to have the output of an arbitrary tensor `t` swapped to CPU memory?*]

Overall I think it would be cool, if this could be pinged to the right person at the tensorflow team. I feel that the large memory usage is the [biggest down-side of tensorflow](https://github.com/tensorflow/tensorflow/issues/492). In all models I have debugged (including standard CNN ones), the majority of memory is blocked by tensors being kept in memory for gradient computation. 

We know two solutions for this. Namely `recomputation` of some tensors during back-prop (i.e. via node mirroring) or `memory swapping`. Both can be implemented in pure python. So I don't see much reason for not tackling this. This kind of optimization will reduce the memory requirement of all very deep models (>= 16 layer) by an order of magnitude.

**Comment by MarvinTeichmann** [comment]
Thanks for the very fast reply, @prb12. I am very happy to learn about existing ways to recompute gradients instead of keeping them. This will allow me to use Tensorflow for my next CRF based project.

I will take a closer look at defun and how it might be used for gradient recomputation after I return from my conference. I am very happy about alternative suggestions in the meantime.

**Comment by prb12** [comment]
One other way to temporarily override gradient code is to use `gradient_override_map`  
https://github.com/tensorflow/tensorflow/blob/2cfbd3347b943a389bb688125c2a90095b7735b5/tensorflow/python/framework/ops.py#L3773

```python
 @tf.RegisterGradient("CustomSquare")
   def _custom_square_grad(op, grad):
   # ...
    
 with tf.Graph().as_default() as g:
   c = tf.constant(5.0)
   s_1 = tf.square(c)  # Uses the default gradient for tf.square.
   with g.gradient_override_map({"Square": "CustomSquare"}):
     s_2 = tf.square(s_2)  # Uses _custom_square_grad to compute the gradient of s_2.
```
Since the custom gradient function is passed the original `Operation`, it can recompute the op from the same inputs.

The main reason this doesn't get used for recomputation is that this mechanism is done at an op granularity, whereas most people probably want to do recomputation at a layer granularity.

The `Defun` trick is basically the same thing but applied to a larger subgraph.

**Comment by ebrevdo** [comment]
We may be able to add a decorator/wrapper that just does this Defun trick.
Call it "@recompute_on_gradient" or something.

**Comment by prb12** [comment]
> [Btw. could someone point me to the line where swap_memory is implemented? 

The `swap_memory` flags get passed through to `tf.while_loop` and into the [`WhileContext`](https://github.com/tensorflow/tensorflow/blob/cf7c008ab150ac8e5edb3ed053d38b2919699796/tensorflow/python/ops/control_flow_ops.py#L1907),  which pass flags to the stack related ops [here](https://github.com/tensorflow/tensorflow/blob/cf7c008ab150ac8e5edb3ed053d38b2919699796/tensorflow/python/ops/control_flow_ops.py#L877-L878). 

`StackPush` operations have an [attr](https://github.com/tensorflow/tensorflow/blob/cf7c008ab150ac8e5edb3ed053d38b2919699796/tensorflow/core/ops/data_flow_ops.cc#L1132-L1149) to enable the functionality, and the actual swapping happens [here](https://github.com/tensorflow/tensorflow/blob/cf7c008ab150ac8e5edb3ed053d38b2919699796/tensorflow/core/kernels/stack_ops.cc#L222-L259).

**Comment by yaroslavvb** [comment]
Getting Chen's performance is too complicated for gradient decorator -- for a uniform chain of length n, you want to insert "checkpoints" every sqrt(n) nodes, which doesn't necessarily correspond to number of gradient calls. This kind of mirroring is indeed possible by using contrib.graph_editor to clone nodes. This enables a node execution order which uses less memory, and you can use control dependencies to force TensorFlow to use this schedule. In other words, if you have a node like "conv-forward", you could duplicate it and then disconnect one of them from the gradient. Since the original node is not connected to gradient computation, its value can be forgotten at the stage of forward prop.

TensorFlow default strategy will execute nodes as soon as possible, which is the opposite of what you want to do you in your scenario. Here's a [utility](https://github.com/yaroslavvb/stuff/tree/master/linearize) that forces TensorFlow to execute nodes as late as possible, breaking ties by using estimate of memory usage.

Forgetting intermediate results to fit in memory budget is actually a pretty interesting general problem, studied in compiler literature as "one-shot pebbling". It can be solved exactly for chains, and I suspect can also be solved exactly for typical neural networks, since they tend to have small treewidth (ie, you can pick size-k subgraphs, and treat them as nodes with 2^k states such that resulting "merged" graph is a chain)).

**Comment by ebrevdo** [comment]
Not sure I fully understand your checkpoints model, Yaroslav.  How is it
different from just using the Defun every sqrt(n) nodes - and the rest of
the time using the standard calls?

**Comment by yaroslavvb** [comment]
@ebrevdo manually wrapping sqrt(n) nodes into defun would do the trick, I just didn't get how the decorator would do this.

There's some ambiguity in docs on whether Defun recomputes the values. There's `noinline=True` option to Defun, although when I tried on 1.1, it worked without having to set that attribute.

PS, swapping stuff to main memory is not that great for O(n) operations. I benchmarked it on TitanX, and recomputing large `Mul` from inputs on GPU was 7x faster than fetching it from main memory. Recomputing `Concat` was 10x faster

**Comment by ebrevdo** [comment]
To do fw recomputation in the gradient with a defun you would do something
like:

def fw(x1, x2):  # x1 and x2 are tensors
  ...
  return y1, y2  # y1 and y2 are tensors

def recalc_fw_grad(fw_op, dy1, dy2):
  x1, x2 = fw_op.inputs
  # DON'T USE THESE: # y1, y2 = fw_op.outputs
  y1, y2 = fw(x1, x2)
  return tf.gradients(ys=[y1, y2], xs=[x1, x2], grad_ys=[dy1, dy2])

@Defun(python_grad_func=recalc_fw_grad, shape_func=...)
def fw_with_recalc_fw_grad(x1, x2):
  return fw(x1, x2)

**Comment by ebrevdo** [comment]
You probably also want to add noinline=True to the Defun constructor.

**Comment by alexlee-gk** [comment]
Hi, I'd like to reopen this discussion on whether there is a better way to do this. I followed @ebrevdo suggestion and implemented a `recompute_on_gradient` decorator that uses `Defun`. I have included the decorator and sample usage below. It works as desired, but it has some major limitations:
1. I couldn't find a good way to pass in function parameters, e.g. `reduction_dim` in the example below.
2. The tensors being passed to the grad function are of unknown shape, so you can't figure out certain properties of the inputs, e.g. `broadcast_dim_a` and `broadcast_dim_b` in the example below.

For 1., I tried using a closure to pass in the parameters, but that significantly slows down the execution. I also tried using `tf.name_scope`, but that is not that useful since the grad function doesn't get the `op`.

In general, it would be really useful to have a way of defining a subgraph of ops as a single op so that you can define a gradient for that subgraph as a whole, and thus preventing tensorflow from storing the intermediate tensors (which can be really huge) of that subgraph for the backward pass.

I have looked into `gradient_override_map` and `tf.py_func`, but they don't seem to be applicable here. The former only seems to work for individual ops that have been defined in C++. The latter does group a subgraph of ops into a single one, however this is for numpy functions. What's needed is something like `tf.tf_func` for a subgraph of tensorflow ops.

```
def recompute_on_gradient(func_name, grad_func=None):
    """
    Assumes func takes in multiple inputs and returns a single output.
    """
    def _recompute_on_gradient(func):
        def _shape_func(op):
            try:
                return [output.shape for output in func(*op.inputs)]
            except TypeError:
                return [func(*op.inputs).shape]

        if grad_func is None:
            def _grad_func(*args):
                inputs = list(args[:-1])
                grad = args[-1]
                output = func(*inputs)
                return tf.gradients(ys=[output], xs=inputs, grad_ys=[grad])
        else:
            _grad_func = grad_func

        _grad_func = function.Defun(func_name=func_name + "_grad",
                                    noinline=True)(_grad_func)

        return function.Defun(grad_func=_grad_func,
                              shape_func=_shape_func,
                              func_name=func_name,
                              noinline=True)(func)
    return _recompute_on_gradient


def _multiply_sum_grad(a, b, grad_c):
    reduction_dim = -2
    broadcast_dim_a = -1
    broadcast_dim_b = 0

    grad_c = tf.expand_dims(grad_c, axis=reduction_dim)
    grad_a = grad_c * b
    grad_b = grad_c * a
    # sum over broadcast dimensions
    grad_a = tf.reduce_sum(grad_a, axis=broadcast_dim_a, keep_dims=True)
    grad_b = tf.reduce_sum(grad_b, axis=broadcast_dim_b, keep_dims=True)
    return grad_a, grad_b


@recompute_on_gradient("multiply_sum", _multiply_sum_grad)
def multiply_sum(a, b):
    reduction_dim = -2
    return tf.reduce_sum(a * b, axis=reduction_dim)


@recompute_on_gradient("multiply_sum_no_grad")
def multiply_sum_no_grad(a, b):
    reduction_dim = -2
    return tf.reduce_sum(a * b, axis=reduction_dim)


time = 3
batch = 16
in_height = 64
in_width = 64
in_channels = 1
kernel_size = [17, 17]
filters = 4
patches = tf.get_variable('a', [time, batch, in_height, in_width, kernel_size[0] * kernel_size[1] * in_channels])
kernel = tf.get_variable('b', [batch, in_height, in_width, kernel_size[0], kernel_size[1], in_channels, filters])
kernel_reshaped = tf.reshape(kernel, [batch, in_height, in_width, kernel_size[0] * kernel_size[1] * in_channels, filters])

output = tf.reduce_sum(patches[..., None] * kernel_reshaped[None, ...], axis=-2)  # uses more than 3GB to compute the gradient
output_mulsum = multiply_sum(patches[..., None], kernel_reshaped[None, ...])
output_mulsum_no_grad = multiply_sum_no_grad(patches[..., None], kernel_reshaped[None, ...])
```

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and the `awaiting tensorflower` label was assigned. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and the `awaiting tensorflower` label was assigned. Please update the label and/or status accordingly.

**Comment by alexlee-gk** [comment]
That looks great, thanks!

**Comment by tatatodd** [comment]
Very cool @yaroslavvb!

@MarvinTeichmann does Yaroslav's package resolve this issue?

**Comment by tensorflowbutler** [comment]
Nagging Awaiting Response: It has been 14 days with no activityand the `awaiting response` label was assigned. Is this still an issue?

**Comment by tensorflowbutler** [comment]
Nagging Awaiting Response: It has been 14 days with no activityand the `awaiting response` label was assigned. Is this still an issue?

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and the `awaiting response` label was assigned. Is this still an issue?

**Comment by reedwm** [comment]
/CC @prb12 any updates on this? @yaroslavvb package is very interesting. Should something similar be in TensorFlow itself?

**Comment by yaroslavvb** [comment]
@reedwm I think Grappler is eventually supposed to take over these things (@allenlavoie )

**Comment by allenlavoie** [comment]
There's also the [recompute_grad decorator](https://www.tensorflow.org/api_docs/python/tf/contrib/layers/recompute_grad) now. And Grappler does some swapping to host memory by default. But yes, it'd be nice if Grappler's memory optimizer did checkpointing automatically keeping device memory constraints in mind. (But unfortunately I have no immediate plans to work on it.)

**Comment by reedwm** [comment]
I'll assign to you @allenlavoie and mark as "awaiting tensorflow", which now means you currently aren't working on it.

**Comment by allenlavoie** [comment]
I think it's fine as "contributions welcome". I'll grab the bug if I start working on it. Happy to discuss if anyone wants to take it first.

**Comment by joeyearsley** [comment]
Maybe I'm doing something wrong, but the recompute grad decorator isn't seemingly saving myself any memory. I'm using it naively on every `tf.layers.conv2d` like so:

```
@recompute_grad
def _x(x):
    return tf.layers.conv2d(x, ...)
```
However, it uses the same amount of memory as standard TF 1.10 . 

Also, I used to have gradient_checkpointing saving myself memory (back in april) however it now doesn't save any memory either.

@allenlavoie Any ideas why?

**Comment by allenlavoie** [comment]
@joeyearsley recompute_grad needs to catch an intermediate activation which would otherwise feed into a gradient op in order to save memory. `tf.layers.conv2d` has a conv op and maybe a ReLU, and ReLU automatically recomputes, so if the model is just a sequence of `_x` calls then it's already saving only one activation per `_x`. To save memory, you could wrap a sequence of `_x` calls in `recompute_grad` (but include too many and the recomputation of the gradient will take as much memory as the model would have otherwise). If there's a batch norm layer I'd include that in the decorated function, since it isn't automatically recomputed like ReLU.

**Comment by sushreebarsa** [comment]
@MarvinTeichmann An lternative approaches for memory reduction would be as follows;

1. Gradient checkpointing: Saves intermediate activations during backpropagation, allowing reuse and reducing GPU memory footprint.
2. Mixed precision training: Uses lower precision formats like fp16 for calculations without significantly impacting accuracy.
Thank you!
